File: codex-rs/core/src/conversation_manager.rs
use crate::AuthManager;
#[cfg(any(test, feature = "test-support"))]
use crate::CodexAuth;
#[cfg(any(test, feature = "test-support"))]
use crate::ModelProviderInfo;
use crate::codex::Codex;
use crate::codex::CodexSpawnOk;
use crate::codex::INITIAL_SUBMIT_ID;
use crate::codex_conversation::CodexConversation;
use crate::config::Config;
use crate::error::CodexErr;
use crate::error::Result as CodexResult;
use crate::history_truncation;
use crate::models_manager::manager::ModelsManager;
use crate::protocol::Event;
use crate::protocol::EventMsg;
use crate::protocol::SessionConfiguredEvent;
use crate::rollout::RolloutRecorder;
use crate::skills::SkillsManager;
use codex_protocol::ConversationId;
use codex_protocol::openai_models::ModelPreset;
use codex_protocol::protocol::InitialHistory;
use codex_protocol::protocol::SessionSource;
use std::collections::HashMap;
use std::path::PathBuf;
use std::sync::Arc;
#[cfg(any(test, feature = "test-support"))]
use tempfile::TempDir;
use tokio::sync::RwLock;

/// Represents a newly created Codex conversation, including the first event
/// (which is [`EventMsg::SessionConfigured`]).
pub struct NewConversation {
    pub conversation_id: ConversationId,
    pub conversation: Arc<CodexConversation>,
    pub session_configured: SessionConfiguredEvent,
}

/// [`ConversationManager`] is responsible for creating conversations and
/// maintaining them in memory.
pub struct ConversationManager {
    conversations: Arc<RwLock<HashMap<ConversationId, Arc<CodexConversation>>>>,
    auth_manager: Arc<AuthManager>,
    models_manager: Arc<ModelsManager>,
    skills_manager: Arc<SkillsManager>,
    session_source: SessionSource,
    #[cfg(any(test, feature = "test-support"))]
    _test_codex_home_guard: Option<TempDir>,
}

impl ConversationManager {
    pub fn new(auth_manager: Arc<AuthManager>, session_source: SessionSource) -> Self {
        let skills_manager = Arc::new(SkillsManager::new(auth_manager.codex_home().to_path_buf()));
        Self {
            conversations: Arc::new(RwLock::new(HashMap::new())),
            auth_manager: auth_manager.clone(),
            session_source,
            models_manager: Arc::new(ModelsManager::new(auth_manager)),
            skills_manager,
            #[cfg(any(test, feature = "test-support"))]
            _test_codex_home_guard: None,
        }
    }

    #[cfg(any(test, feature = "test-support"))]
    /// Construct with a dummy AuthManager containing the provided CodexAuth.
    /// Used for integration tests: should not be used by ordinary business logic.
    pub fn with_models_provider(auth: CodexAuth, provider: ModelProviderInfo) -> Self {
        let temp_dir = tempfile::tempdir().unwrap_or_else(|err| panic!("temp codex home: {err}"));
        let codex_home = temp_dir.path().to_path_buf();
        let mut manager = Self::with_models_provider_and_home(auth, provider, codex_home);
        manager._test_codex_home_guard = Some(temp_dir);
        manager
    }

    #[cfg(any(test, feature = "test-support"))]
    /// Construct with a dummy AuthManager containing the provided CodexAuth and codex home.
    /// Used for integration tests: should not be used by ordinary business logic.
    pub fn with_models_provider_and_home(
        auth: CodexAuth,
        provider: ModelProviderInfo,
        codex_home: PathBuf,
    ) -> Self {
        let auth_manager = crate::AuthManager::from_auth_for_testing_with_home(auth, codex_home);
        let skills_manager = Arc::new(SkillsManager::new(auth_manager.codex_home().to_path_buf()));
        Self {
            conversations: Arc::new(RwLock::new(HashMap::new())),
            auth_manager: auth_manager.clone(),
            session_source: SessionSource::Exec,
            models_manager: Arc::new(ModelsManager::with_provider(auth_manager, provider)),
            skills_manager,
            _test_codex_home_guard: None,
        }
    }

    pub fn session_source(&self) -> SessionSource {
        self.session_source.clone()
    }

    pub fn skills_manager(&self) -> Arc<SkillsManager> {
        self.skills_manager.clone()
    }

    pub async fn new_conversation(&self, config: Config) -> CodexResult<NewConversation> {
        self.spawn_conversation(
            config,
            self.auth_manager.clone(),
            self.models_manager.clone(),
        )
        .await
    }

    async fn spawn_conversation(
        &self,
        config: Config,
        auth_manager: Arc<AuthManager>,
        models_manager: Arc<ModelsManager>,
    ) -> CodexResult<NewConversation> {
        let CodexSpawnOk {
            codex,
            conversation_id,
        } = Codex::spawn(
            config,
            auth_manager,
            models_manager,
            self.skills_manager.clone(),
            InitialHistory::New,
            self.session_source.clone(),
        )
        .await?;
        self.finalize_spawn(codex, conversation_id).await
    }

    async fn finalize_spawn(
        &self,
        codex: Codex,
        conversation_id: ConversationId,
    ) -> CodexResult<NewConversation> {
        // The first event must be `SessionInitialized`. Validate and forward it
        // to the caller so that they can display it in the conversation
        // history.
        let event = codex.next_event().await?;
        let session_configured = match event {
            Event {
                id,
                msg: EventMsg::SessionConfigured(session_configured),
            } if id == INITIAL_SUBMIT_ID => session_configured,
            _ => {
                return Err(CodexErr::SessionConfiguredNotFirstEvent);
            }
        };

        let conversation = Arc::new(CodexConversation::new(
            codex,
            session_configured.rollout_path.clone(),
        ));
        self.conversations
            .write()
            .await
            .insert(conversation_id, conversation.clone());

        Ok(NewConversation {
            conversation_id,
            conversation,
            session_configured,
        })
    }

    pub async fn get_conversation(
        &self,
        conversation_id: ConversationId,
    ) -> CodexResult<Arc<CodexConversation>> {
        let conversations = self.conversations.read().await;
        conversations
            .get(&conversation_id)
            .cloned()
            .ok_or_else(|| CodexErr::ConversationNotFound(conversation_id))
    }

    pub async fn resume_conversation_from_rollout(
        &self,
        config: Config,
        rollout_path: PathBuf,
        auth_manager: Arc<AuthManager>,
    ) -> CodexResult<NewConversation> {
        let initial_history = RolloutRecorder::get_rollout_history(&rollout_path).await?;
        self.resume_conversation_with_history(config, initial_history, auth_manager)
            .await
    }

    pub async fn resume_conversation_with_history(
        &self,
        config: Config,
        initial_history: InitialHistory,
        auth_manager: Arc<AuthManager>,
    ) -> CodexResult<NewConversation> {
        let CodexSpawnOk {
            codex,
            conversation_id,
        } = Codex::spawn(
            config,
            auth_manager,
            self.models_manager.clone(),
            self.skills_manager.clone(),
            initial_history,
            self.session_source.clone(),
        )
        .await?;
        self.finalize_spawn(codex, conversation_id).await
    }

    /// Removes the conversation from the manager's internal map, though the
    /// conversation is stored as `Arc<CodexConversation>`, it is possible that
    /// other references to it exist elsewhere. Returns the conversation if the
    /// conversation was found and removed.
    pub async fn remove_conversation(
        &self,
        conversation_id: &ConversationId,
    ) -> Option<Arc<CodexConversation>> {
        self.conversations.write().await.remove(conversation_id)
    }

    /// Fork an existing conversation by taking messages up to the given position
    /// (not including the message at the given position) and starting a new
    /// conversation with identical configuration (unless overridden by the
    /// caller's `config`). The new conversation will have a fresh id.
    pub async fn fork_conversation(
        &self,
        nth_user_message: usize,
        config: Config,
        path: PathBuf,
    ) -> CodexResult<NewConversation> {
        // Compute the prefix up to the cut point.
        let history = RolloutRecorder::get_rollout_history(&path).await?;
        let rollout_items = history.get_rollout_items();
        let truncated = history_truncation::truncate_rollout_before_nth_user_message_from_start(
            &rollout_items,
            nth_user_message,
        );
        let history = if truncated.is_empty() {
            InitialHistory::New
        } else {
            InitialHistory::Forked(truncated)
        };

        // Spawn a new conversation with the computed initial history.
        let auth_manager = self.auth_manager.clone();
        let CodexSpawnOk {
            codex,
            conversation_id,
        } = Codex::spawn(
            config,
            auth_manager,
            self.models_manager.clone(),
            self.skills_manager.clone(),
            history,
            self.session_source.clone(),
        )
        .await?;

        self.finalize_spawn(codex, conversation_id).await
    }

    pub async fn list_models(&self, config: &Config) -> Vec<ModelPreset> {
        self.models_manager.list_models(config).await
    }

    pub fn get_models_manager(&self) -> Arc<ModelsManager> {
        self.models_manager.clone()
    }
}

File: src/nativeMain/kotlin/ai/solace/coder/core/session/ConversationManager.kt
// port-lint: source codex-rs/core/src/conversation_manager.rs
package ai.solace.coder.core.session

import ai.solace.coder.client.auth.AuthManager
import ai.solace.coder.core.config.Config
import ai.solace.coder.core.error.CodexError
import ai.solace.coder.core.error.CodexResult
import ai.solace.coder.protocol.ConversationId
import ai.solace.coder.protocol.EventMsg
import ai.solace.coder.protocol.SessionConfiguredEvent
import ai.solace.coder.protocol.SessionSource
import ai.solace.coder.protocol.ResponseItem
import ai.solace.coder.protocol.ContentItem
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * Represents a newly created Codex conversation, including the first event
 * (which is SessionConfigured).
 */
data class NewConversation(
    val conversationId: ConversationId,
    val conversation: CodexConversation,
    val sessionConfigured: SessionConfiguredEvent
)

/**
 * Wrapper around a Codex instance that represents an active conversation.
 */
class CodexConversation(
    val codex: Codex,
    val rolloutPath: String?
)

/**
 * The initial submit ID used for the session configuration event.
 */
const val INITIAL_SUBMIT_ID: String = "initial"

/**
 * Initial history for conversation creation.
 */
sealed class InitialHistory {
    /** Brand new conversation. */
    data object New : InitialHistory()
    /** Forked from existing conversation with items. */
    data class Forked(val items: List<RolloutItem>) : InitialHistory()

    fun getRolloutItems(): List<RolloutItem> = when (this) {
        is New -> emptyList()
        is Forked -> items
    }
}

/**
 * An item in a conversation rollout.
 */
sealed class RolloutItem {
    data class ResponseItemWrapper(val item: ResponseItem) : RolloutItem()
}

/**
 * ConversationManager is responsible for creating conversations and
 * maintaining them in memory.
 *
 * Ported from Rust codex-rs/core/src/conversation_manager.rs
 */
class ConversationManager(
    private val authManager: AuthManager,
    private val sessionSource: SessionSource
) {
    private val conversations = mutableMapOf<ConversationId, CodexConversation>()
    private val mutex = Mutex()

    fun sessionSource(): SessionSource = sessionSource

    /**
     * Create a new conversation with the given config.
     */
    suspend fun newConversation(config: Config): CodexResult<NewConversation> {
        return spawnConversation(config, authManager)
    }

    /**
     * Get an existing conversation by its ID.
     */
    suspend fun getConversation(conversationId: ConversationId): CodexResult<CodexConversation> {
        return mutex.withLock {
            val conversation = conversations[conversationId]
            if (conversation != null) {
                CodexResult.success(conversation)
            } else {
                CodexResult.failure(CodexError.ConversationNotFound(conversationId.toString()))
            }
        }
    }

    /**
     * Remove a conversation from the manager.
     */
    suspend fun removeConversation(conversationId: ConversationId): CodexConversation? {
        return mutex.withLock {
            conversations.remove(conversationId)
        }
    }

    /**
     * Resume a conversation with the given initial history.
     */
    suspend fun resumeConversationWithHistory(
        config: Config,
        initialHistory: InitialHistory,
        authManager: AuthManager
    ): CodexResult<NewConversation> {
        // TODO: Spawn with initial history once Codex.spawn supports it
        return spawnConversation(config, authManager)
    }

    /**
     * Fork an existing conversation by taking messages up to the given position.
     */
    suspend fun forkConversation(
        nthUserMessage: Int,
        config: Config,
        rolloutPath: String
    ): CodexResult<NewConversation> {
        // TODO: Load rollout history and truncate
        return spawnConversation(config, authManager)
    }

    private suspend fun spawnConversation(
        config: Config,
        authManager: AuthManager
    ): CodexResult<NewConversation> {
        // TODO: Call Codex.spawn() once it's fully implemented
        // The flow is:
        // 1. Codex.spawn(config, authManager, InitialHistory.New, sessionSource)
        // 2. Wait for first SessionConfigured event
        // 3. Create CodexConversation and register it
        return CodexResult.failure(
            CodexError.NotImplemented("ConversationManager.spawnConversation requires full Codex.spawn implementation")
        )
    }

    /**
     * Finalize a spawned conversation by validating the first event
     * and registering the conversation.
     */
    internal suspend fun finalizeSpawn(
        codex: Codex,
        conversationId: ConversationId,
        rolloutPath: String?
    ): CodexResult<NewConversation> {
        // The first event must be SessionConfigured. Validate and forward it
        // to the caller so they can display it in the conversation history.
        val eventResult = codex.nextEvent()
        val event = eventResult.getOrNull()
            ?: return CodexResult.failure(CodexError.SessionConfiguredNotFirstEvent)

        val sessionConfigured = when (val msg = event.msg) {
            is EventMsg.SessionConfigured -> msg.payload
            else -> return CodexResult.failure(CodexError.SessionConfiguredNotFirstEvent)
        }

        val conversation = CodexConversation(codex, rolloutPath)

        mutex.withLock {
            conversations[conversationId] = conversation
        }

        return CodexResult.success(
            NewConversation(
                conversationId = conversationId,
                conversation = conversation,
                sessionConfigured = sessionConfigured
            )
        )
    }
}

/**
 * Return a prefix of items obtained by cutting strictly before the nth user
 * message (0-based) and all items that follow it.
 */
fun truncateBeforeNthUserMessage(history: InitialHistory, n: Int): InitialHistory {
    val items = history.getRolloutItems()

    // Find indices of user message inputs in rollout order
    val userPositions = mutableListOf<Int>()
    for ((idx, item) in items.withIndex()) {
        if (item is RolloutItem.ResponseItemWrapper) {
            val responseItem = item.item
            if (responseItem is ResponseItem.Message && responseItem.role == "user") {
                userPositions.add(idx)
            }
        }
    }

    // If fewer than or equal to n user messages exist, treat as empty
    if (userPositions.size <= n) {
        return InitialHistory.New
    }

    // Cut strictly before the nth user message
    val cutIdx = userPositions[n]
    val rolled = items.take(cutIdx)

    return if (rolled.isEmpty()) {
        InitialHistory.New
    } else {
        InitialHistory.Forked(rolled)
    }
}
